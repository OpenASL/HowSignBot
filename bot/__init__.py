__version__ = "21.12.1"
