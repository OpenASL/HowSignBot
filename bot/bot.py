import asyncio
import itertools
import logging

import disnake
from disnake import ApplicationCommandInteraction
from disnake.ext import commands

from . import settings

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX


intents = disnake.Intents.default()
intents.typing = False
intents.presences = False
intents.bans = False
intents.integrations = False
intents.webhooks = False
intents.invites = False
intents.members = True
intents.messages = True
bot = commands.Bot(
    command_prefix=COMMAND_PREFIX,
    case_insensitive=True,
    owner_id=settings.OWNER_ID,
    intents=intents,
    sync_commands_debug=settings.DEBUG,
    reload=settings.DEBUG,
    test_guilds=settings.TEST_GUILDS or None,
)

PRESENCE_COMMANDS = (
    "/sign",
    "/handshape",
    "/catchphrase | /codenames",
    "/schedule new",
    "/watch2gether",
    "/sentence | /idiom",
    "/feedback",
)
CHANGE_PRESENCE_EVERY = 120  # seconds


async def cycle_presence():
    for command in itertools.cycle(PRESENCE_COMMANDS):
        activity = disnake.Activity(
            name=command,
            type=disnake.ActivityType.watching,
        )
        logger.debug(f"changing presence to show command: {command}")
        await bot.change_presence(activity=activity)
        await asyncio.sleep(CHANGE_PRESENCE_EVERY)


@bot.event
async def on_ready():
    bot.loop.create_task(cycle_presence())


@bot.event
async def on_command_error(ctx, error):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await ctx.send(error)
    else:
        logger.error(f"unhandled exception from command: {ctx.command}", exc_info=error)


@bot.event
async def on_slash_command_error(inter: ApplicationCommandInteraction, error: Exception):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await inter.send(error)
    else:
        logger.error(
            f"unhandled exception from command: {inter.application_command.name!r}",
            exc_info=error,
        )


async def set_default_presence():
    activity = disnake.Activity(
        name="/sign",
        type=disnake.ActivityType.watching,
    )
    await bot.change_presence(activity=activity)
