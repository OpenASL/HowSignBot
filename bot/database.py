from __future__ import annotations

import datetime as dt
import logging
from typing import Iterator, Mapping, Sequence

import databases
import nanoid
import pytz
import sqlalchemy as sa
from disnake import Member
from pytz.tzinfo import StaticTzInfo
from sqlalchemy import sql
from sqlalchemy.dialects.postgresql import BIGINT
from sqlalchemy.dialects.postgresql import TIMESTAMP as _TIMESTAMP
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.sql.schema import ForeignKey

from . import settings

metadata = sa.MetaData()
NULL = sql.null()

logger = logging.getLogger(__name__)


def now():
    return dt.datetime.now(dt.timezone.utc)


class TimeZone(sa.TypeDecorator):
    impl = sa.Text

    def process_bind_param(self, value, dialect):
        if value is None:
            return None
        return str(value)

    def process_result_value(self, value, dialect):
        if value is None:
            return None
        return pytz.timezone(value)


class TIMESTAMP(sa.TypeDecorator):
    impl = _TIMESTAMP(timezone=True)


def created_at_column(name="created_at", **kwargs):
    return sa.Column(name, TIMESTAMP, nullable=False, default=now, **kwargs)


NANOID_ALPHABET = "23456789abcdefghijkmnopqrstuvwxyz"
NANOID_SIZE = 5


def generate_nanoid() -> str:
    return nanoid.generate(alphabet=NANOID_ALPHABET, size=NANOID_SIZE)


# -----------------------------------------------------------------------------

guild_settings = sa.Table(
    "guild_settings",
    metadata,
    sa.Column("guild_id", BIGINT, primary_key=True, doc="Discord guild ID"),
    sa.Column("name", sa.Text, doc="Name of the guild, for readability in clients"),
    sa.Column(
        "schedule_sheet_key",
        sa.Text,
        doc="Google sheet key for the guild's practice schedule",
    ),
    sa.Column(
        "daily_message_channel_id",
        BIGINT,
        doc="Discord channel ID where to post daily schedule message.",
    ),
    sa.Column(
        "include_practice_schedule",
        sa.Boolean,
        server_default=sql.true(),
        nullable=False,
    ),
    sa.Column(
        "include_handshape_of_the_day",
        sa.Boolean,
        server_default=sql.false(),
        nullable=False,
    ),
    sa.Column(
        "include_topics_of_the_day",
        sa.Boolean,
        server_default=sql.false(),
        nullable=False,
    ),
    sa.Column(
        "include_clthat",
        sa.Boolean,
        server_default=sql.false(),
        nullable=False,
    ),
)

guild_announcements = sa.Table(
    "guild_announcements",
    metadata,
    sa.Column(
        "guild_id",
        ForeignKey(guild_settings.c.guild_id, ondelete="CASCADE"),
        primary_key=True,
    ),
    sa.Column("title", sa.Text, nullable=False, doc="Announcement title"),
    sa.Column("description", sa.Text, nullable=False, doc="Announcement description"),
    sa.Column(
        "is_active",
        sa.Boolean,
        nullable=False,
        server_default=sql.true(),
        doc="Whether to display the announcement in daily messages",
    ),
    created_at_column(),
)

user_settings = sa.Table(
    "user_settings",
    metadata,
    sa.Column("user_id", BIGINT, primary_key=True, doc="Discord user ID"),
    sa.Column("timezone", TimeZone),
)

sign_cafe_intros = sa.Table(
    "sign_cafe_intros",
    metadata,
    sa.Column("message_id", BIGINT, primary_key=True, doc="Discord message ID"),
    sa.Column("user_id", BIGINT, index=True, nullable=False, doc="Discord user ID"),
    sa.Column("posted_at", TIMESTAMP, nullable=False),
    created_at_column(),
)

sign_cafe_members = sa.Table(
    "sign_cafe_members",
    metadata,
    sa.Column("user_id", BIGINT, primary_key=True, doc="Discord user ID"),
    sa.Column("joined_at", TIMESTAMP, nullable=False),
    sa.Column("is_active", sa.Boolean, nullable=False, server_default=sql.false()),
    sa.Column("roles", sa.Text, nullable=True),
    sa.Column(
        "has_acknowledged_rules", sa.Boolean, nullable=False, server_default=sql.false()
    ),
    created_at_column(),
)

zoom_meetings = sa.Table(
    "zoom_meetings",
    metadata,
    sa.Column("meeting_id", BIGINT, primary_key=True, doc="Meeting ID issued by Zoom"),
    sa.Column("zoom_user", sa.Text, nullable=False),
    sa.Column("join_url", sa.Text, nullable=False),
    sa.Column("passcode", sa.Text, nullable=False),
    sa.Column("topic", sa.Text, nullable=False),
    sa.Column("host_id", sa.Text, doc="Last cached zoom ID of the host"),
    sa.Column("setup_at", TIMESTAMP),
    created_at_column(),
)

zzzzoom_meetings = sa.Table(
    "zzzzoom_meetings",
    metadata,
    sa.Column("id", sa.String(32), primary_key=True),
    sa.Column(
        "meeting_id",
        ForeignKey(zoom_meetings.c.meeting_id, ondelete="CASCADE"),
        index=True,
    ),
    created_at_column(),
)

zoom_messages = sa.Table(
    "zoom_messages",
    metadata,
    sa.Column("message_id", BIGINT, primary_key=True, doc="Discord message ID"),
    sa.Column("channel_id", BIGINT, nullable=False, doc="Discord channel ID"),
    sa.Column("meeting_id", ForeignKey(zoom_meetings.c.meeting_id, ondelete="CASCADE")),
    created_at_column(),
)

zoom_participants = sa.Table(
    "zoom_participants",
    metadata,
    sa.Column(
        "meeting_id",
        ForeignKey(zoom_meetings.c.meeting_id, ondelete="CASCADE"),
        primary_key=True,
    ),
    sa.Column("name", sa.Text, primary_key=True),
    sa.Column("zoom_id", sa.Text, doc="Zoom user ID (null for non-registered users)"),
    sa.Column("email", sa.Text, doc="Zoom email (null for non-registered users)"),
    sa.Column(
        "joined_at",
        TIMESTAMP,
        doc="Join time from webhook payload. Unlike created_at, this will be updated when a participant is moved to and from a breakout room.",
    ),
    created_at_column(),
)

topics = sa.Table(
    "topics",
    metadata,
    sa.Column("content", sa.Text, primary_key=True),
    sa.Column("last_synced_at", TIMESTAMP),
)

scheduled_events = sa.Table(
    "scheduled_events",
    metadata,
    sa.Column("event_id", BIGINT, primary_key=True, doc="Discord event ID"),
    sa.Column(
        "created_by",
        BIGINT,
        index=True,
        doc="Discord user ID for the user who created the event through the bot",
    ),
    created_at_column(),
)

# -----------------------------------------------------------------------------


class Store:
    metadata = metadata

    def __init__(
        self,
        database_url: str | databases.DatabaseURL,
        *,
        force_rollback: bool = False,
    ):
        self.db = databases.Database(database_url, force_rollback=force_rollback)

    def connect(self):
        return self.db.connect()

    def disconnect(self):
        return self.db.disconnect()

    def transaction(self):
        return self.db.transaction()

    async def set_user_timezone(self, user_id: int, timezone: dt.tzinfo | None):
        logger.info(f"setting timezone for user_id {user_id}")
        stmt = insert(user_settings).values(user_id=user_id, timezone=timezone)
        stmt = stmt.on_conflict_do_update(
            index_elements=(user_settings.c.user_id,),
            set_=dict(timezone=stmt.excluded.timezone),
        )
        await self.db.execute(stmt)

    async def get_user_timezone(self, user_id: int) -> StaticTzInfo | None:
        logger.info(f"retrieving timezone for user_id {user_id}")
        query = user_settings.select().where(user_settings.c.user_id == user_id)
        return await self.db.fetch_val(query=query, column=user_settings.c.timezone)

    async def get_guild_settings(self, guild_id: int) -> Mapping | None:
        logger.info(f"retrieving guild settings sheet key for guild_id {guild_id}")
        query = guild_settings.select().where(guild_settings.c.guild_id == guild_id)
        return await self.db.fetch_one(query=query)

    async def get_guild_announcements(self, guild_id: int) -> list[Mapping]:
        query = (
            guild_announcements.select()
            .where(
                (guild_announcements.c.guild_id == guild_id)
                & (guild_announcements.c.is_active == sql.true())
            )
            .order_by(guild_announcements.c.created_at.desc())
        )
        return await self.db.fetch_all(query=query)

    async def get_guild_schedule_sheet_key(self, guild_id: int) -> str | None:
        query = guild_settings.select().where(guild_settings.c.guild_id == guild_id)
        return await self.db.fetch_val(
            query=query, column=guild_settings.c.schedule_sheet_key
        )

    async def get_guild_daily_message_channel_id(self, guild_id: int) -> int | None:
        query = guild_settings.select().where(guild_settings.c.guild_id == guild_id)
        return await self.db.fetch_val(
            query=query, column=guild_settings.c.daily_message_channel_id
        )

    async def guild_has_practice_schedule(self, guild_id: int) -> bool:
        select = sa.select(
            (
                sa.exists()
                .where(
                    (guild_settings.c.guild_id == guild_id)
                    & (guild_settings.c.schedule_sheet_key != NULL)
                )
                .label("result"),
            )
        )
        record = await self.db.fetch_one(select)
        if not record:
            return False
        return record["result"]

    async def get_guild_ids_with_practice_schedules(self) -> Iterator[int]:
        all_settings = await self.db.fetch_all(
            guild_settings.select().where(
                guild_settings.c.daily_message_channel_id != NULL
            )
        )
        return (record["guild_id"] for record in all_settings)

    async def get_daily_message_channel_ids(self) -> Iterator[int]:
        all_settings = await self.db.fetch_all(
            guild_settings.select().where(
                guild_settings.c.daily_message_channel_id != NULL
            )
        )
        return (record["daily_message_channel_id"] for record in all_settings)

    # Zoom

    async def create_zoom_meeting(
        self,
        *,
        zoom_user: str,
        meeting_id: int,
        join_url: str,
        passcode: str,
        topic: str,
        set_up: bool,
    ):
        created_at = now()
        stmt = insert(zoom_meetings).values(
            zoom_user=zoom_user,
            meeting_id=meeting_id,
            join_url=join_url,
            passcode=passcode,
            topic=topic,
            # NOTE: need to pass created_at because default=now
            #  doesn't have an effect when using postgresql.insert
            created_at=created_at,
            setup_at=created_at if set_up else None,
        )
        await self.db.execute(stmt)

    async def get_zoom_meeting(self, meeting_id: int) -> Mapping | None:
        query = zoom_meetings.select().where(zoom_meetings.c.meeting_id == meeting_id)
        return await self.db.fetch_one(query=query)

    async def get_latest_pending_zoom_meeting_for_user(
        self, zoom_user: str
    ) -> Mapping | None:
        query = (
            zoom_meetings.select()
            .where(
                (zoom_meetings.c.zoom_user == zoom_user)
                & (zoom_meetings.c.setup_at == NULL)
            )
            .order_by(zoom_meetings.c.created_at.desc())
        )
        return await self.db.fetch_one(query=query)

    async def set_up_zoom_meeting(self, meeting_id: int):
        await self.db.execute(
            zoom_meetings.update()
            .where(zoom_meetings.c.meeting_id == meeting_id)
            .values(setup_at=now())
        )

    async def end_zoom_meeting(self, meeting_id: int):
        await self.db.execute(
            zoom_meetings.delete().where(zoom_meetings.c.meeting_id == meeting_id)
        )
        await self.db.execute(
            zoom_messages.delete().where(zoom_messages.c.meeting_id == meeting_id)
        )
        await self.db.execute(
            zzzzoom_meetings.delete().where(zzzzoom_meetings.c.meeting_id == meeting_id)
        )

    async def set_zoom_meeting_host_id(self, meeting_id: int, *, host_id: str):
        await self.db.execute(
            zoom_meetings.update()
            .where(zoom_meetings.c.meeting_id == meeting_id)
            .values(host_id=host_id)
        )

    async def zoom_meeting_exists(self, meeting_id: int) -> bool:
        select = sa.select(
            (sa.exists().where(zoom_meetings.c.meeting_id == meeting_id).label("result"),)
        )
        record = await self.db.fetch_one(select)
        if not record:
            return False
        return record["result"]

    async def create_zoom_message(
        self, *, meeting_id: int, message_id: int, channel_id: int
    ):
        await self.db.execute(
            insert(zoom_messages).values(
                meeting_id=meeting_id,
                message_id=message_id,
                channel_id=channel_id,
                # NOTE: need to pass created_at because default=now
                #  doesn't have an effect when using postgresql.insert
                created_at=now(),
            )
        )

    async def remove_zoom_message(self, *, message_id: int):
        await self.db.execute(
            zoom_messages.delete().where(zoom_messages.c.message_id == message_id)
        )

    async def get_zoom_message(self, message_id: int) -> Mapping | None:
        return await self.db.fetch_one(
            zoom_messages.select().where(zoom_messages.c.message_id == message_id)
        )

    async def get_zoom_messages(self, meeting_id: int) -> list[Mapping]:
        return await self.db.fetch_all(
            zoom_messages.select().where(zoom_messages.c.meeting_id == meeting_id)
        )

    async def add_zoom_participant(
        self,
        *,
        meeting_id: int,
        name: str,
        zoom_id: str | None,
        email: str | None,
        joined_at: dt.datetime,
    ):
        stmt = insert(zoom_participants).values(
            meeting_id=meeting_id,
            name=name,
            zoom_id=zoom_id,
            email=email,
            joined_at=joined_at,
            # NOTE: need to pass created_at because `default` doesn't execute
            #  when using postgres's insert
            created_at=now(),
        )
        stmt = stmt.on_conflict_do_update(
            index_elements=(zoom_participants.c.meeting_id, zoom_participants.c.name),
            set_=dict(
                zoom_id=stmt.excluded.zoom_id,
                email=stmt.excluded.email,
                joined_at=stmt.excluded.joined_at,
            ),
        )
        await self.db.execute(stmt)

    async def get_zoom_participant(self, *, meeting_id: int, name: str) -> Mapping | None:
        query = zoom_participants.select().where(
            (zoom_participants.c.meeting_id == meeting_id)
            & (zoom_participants.c.name == name)
        )
        return await self.db.fetch_one(query=query)

    async def get_zoom_participants(self, meeting_id: int) -> list[Mapping]:
        return await self.db.fetch_all(
            zoom_participants.select()
            .where(zoom_participants.c.meeting_id == meeting_id)
            .order_by(zoom_participants.c.created_at)
        )

    async def remove_zoom_participant(self, *, meeting_id: int, name: str):
        await self.db.execute(
            zoom_participants.delete().where(
                (zoom_participants.c.meeting_id == meeting_id)
                & (zoom_participants.c.name == name)
            )
        )

    # zzzzoom

    async def create_zzzzoom_meeting(self, *, meeting_id: int):
        created_at = now()
        # TODO: handle id collisions
        stmt = insert(zzzzoom_meetings).values(
            id=generate_nanoid(),
            created_at=created_at,
            meeting_id=meeting_id,
        )
        await self.db.execute(stmt)

    async def get_zzzzoom_meeting(self, id: str):
        query = zzzzoom_meetings.select().where(zzzzoom_meetings.c.id == id)
        return await self.db.fetch_one(query=query)

    async def get_zzzzoom_meeting_for_zoom_meeting(self, meeting_id: int):
        query = zzzzoom_meetings.select().where(
            zzzzoom_meetings.c.meeting_id == meeting_id
        )
        return await self.db.fetch_one(query=query)

    async def zoom_meeting_has_zzzzoom(self, meeting_id: int) -> bool:
        select = sa.select(
            (
                sa.exists()
                .where(zzzzoom_meetings.c.meeting_id == meeting_id)
                .label("result"),
            )
        )
        record = await self.db.fetch_one(select)
        if not record:
            return False
        return record["result"]

    # Topics

    async def save_topics(self, all_topics: Sequence[str]) -> None:
        last_synced_at = now()
        await self.db.execute(topics.delete())
        stmt = insert(topics).values(
            [{"content": topic, "last_synced_at": last_synced_at} for topic in all_topics]
        )
        stmt = stmt.on_conflict_do_update(
            index_elements=(topics.c.content,),
            set_=dict(last_synced_at=stmt.excluded.last_synced_at),
        )
        await self.db.execute(stmt)

    async def get_all_topics(self) -> Sequence[str]:
        all_topics = await self.db.fetch_all(topics.select())
        return [record["content"] for record in all_topics]

    # SIGN_CAFE

    async def add_sign_cafe_intro(
        self,
        *,
        message_id: int,
        user_id: int,
        posted_at: dt.datetime,
    ):
        stmt = insert(sign_cafe_intros).values(
            message_id=message_id,
            user_id=user_id,
            posted_at=posted_at,
            # NOTE: need to pass created_at because `default` doesn't execute
            #  when using postgres's insert
            created_at=now(),
        )
        stmt = stmt.on_conflict_do_update(
            index_elements=(sign_cafe_intros.c.message_id,),
            set_=dict(
                user_id=stmt.excluded.user_id,
                posted_at=stmt.excluded.posted_at,
                created_at=stmt.excluded.created_at,
            ),
        )
        await self.db.execute(stmt)

    async def clear_sign_cafe_members(self):
        await self.db.execute(
            sign_cafe_members.delete().where(sign_cafe_members.c.is_active == sql.false())
        )

    async def clear_sign_cafe_intros(self):
        await self.db.execute(sign_cafe_intros.delete())

    async def get_sign_cafe_member(self, user_id: int) -> Mapping | None:
        return await self.db.fetch_one(
            sign_cafe_members.select().where(sign_cafe_members.c.user_id == user_id)
        )

    async def upsert_sign_cafe_member(self, *, member: Member):
        user_id = member.id
        joined_at = member.joined_at
        role_ids = {role.id for role in member.roles}
        # skip @everyone
        roles_concatenated = "|".join([role.name for role in member.roles[1:]])
        has_acknowledged_rules = settings.SIGN_CAFE_ACKNOWLEDGED_RULES_ROLE_ID in role_ids
        stmt = insert(sign_cafe_members).values(
            user_id=user_id,
            joined_at=joined_at,
            roles=roles_concatenated,
            has_acknowledged_rules=has_acknowledged_rules,
            created_at=now(),
        )
        stmt = stmt.on_conflict_do_update(
            index_elements=(sign_cafe_members.c.user_id,),
            set_=dict(
                joined_at=stmt.excluded.joined_at,
                roles=roles_concatenated,
                has_acknowledged_rules=has_acknowledged_rules,
            ),
        )
        await self.db.execute(stmt)

    async def remove_sign_cafe_member(self, *, user_id: int):
        await self.db.execute(
            sign_cafe_members.delete().where(sign_cafe_members.c.user_id == user_id)
        )

    async def get_sign_cafe_members_without_intro(
        self, since: dt.timedelta
    ) -> list[Mapping]:
        return await self.db.fetch_all(
            sign_cafe_members.select()
            .where(
                (sign_cafe_members.c.is_active == sql.false())
                & (sign_cafe_members.c.has_acknowledged_rules == sql.true())
                & (sign_cafe_members.c.joined_at < (now() - since))
            )
            .select_from(
                sign_cafe_members.outerjoin(
                    sign_cafe_intros,
                    sign_cafe_members.c.user_id == sign_cafe_intros.c.user_id,
                )
            )
            .group_by(sign_cafe_members.c.user_id)
            .having(sa.func.count(sign_cafe_intros.c.user_id) < 1)
            .order_by(sign_cafe_members.c.joined_at)
        )

    async def get_sign_cafe_members_with_no_roles(
        self, leeway: dt.timedelta
    ) -> list[Mapping]:
        return await self.db.fetch_all(
            sign_cafe_members.select()
            .where(
                (sign_cafe_members.c.is_active == sql.false())
                & (sign_cafe_members.c.roles == "")
                & (sign_cafe_members.c.joined_at < (now() - leeway))
            )
            .order_by(sign_cafe_members.c.joined_at)
        )

    async def has_sign_cafe_intro(self, user_id: int) -> bool:
        select = sa.select(
            (sa.exists().where(sign_cafe_intros.c.user_id == user_id).label("result"),)
        )
        record = await self.db.fetch_one(select)
        if not record:
            return False
        return record["result"]

    async def mark_sign_cafe_members_active(self, user_ids: list[int]):
        await self.db.execute(
            sign_cafe_members.update()
            .where(sign_cafe_members.c.user_id.in_(user_ids))
            .values(is_active=True)
        )

    async def mark_sign_cafe_members_inactive(self, user_ids: list[int]):
        await self.db.execute(
            sign_cafe_members.update()
            .where(sign_cafe_members.c.user_id.in_(user_ids))
            .values(is_active=False)
        )

    # Scheduled events

    async def create_scheduled_event(self, *, event_id: int, created_by: int):
        await self.db.execute(
            insert(scheduled_events).values(
                event_id=event_id,
                created_by=created_by,
                # NOTE: need to pass created_at because default=now
                #  doesn't have an effect when using postgresql.insert
                created_at=now(),
            )
        )

    async def get_scheduled_events_for_user(self, user_id: int) -> list[Mapping]:
        query = scheduled_events.select().where(scheduled_events.c.created_by == user_id)
        return await self.db.fetch_all(query=query)

    async def remove_scheduled_event(self, *, event_id: int):
        await self.db.execute(
            scheduled_events.delete().where(scheduled_events.c.event_id == event_id)
        )


store = Store(
    database_url=settings.TEST_DATABASE_URL
    if settings.TESTING
    else settings.DATABASE_URL,
    force_rollback=settings.TESTING,
)
