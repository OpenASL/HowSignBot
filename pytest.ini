[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
