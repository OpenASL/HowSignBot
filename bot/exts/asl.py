import logging
from urllib.parse import quote_plus

import discord
from discord.ext.commands import Cog, Bot, Context, command, errors

import handshapes
from bot import settings
from bot.utils import get_spoiler_text, did_you_mean

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX

SIGN_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for {word_uppercased}]({signingsavvy})
[🌐 **Spread The Sign** - {word_uppercased}]({spread_the_sign})
[📹 **YouGlish** - ASL videos with {word_uppercased}]({youglish})
Share: {howsign}
"""

SIGN_SPOILER_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for ||{word_uppercased}||]({signingsavvy})
[🌐 **Spread The Sign** - ||{word_uppercased}||]({spread_the_sign})
[📹 **YouGlish** - ASL videos with ||{word_uppercased}||]({youglish})
Share: ||{howsign}||
"""

SIGN_HELP = """Look up a word or phrase

If the word or phrase is sent in spoiler text, i.e. enclosed in `||`, the word will also be blacked out in the reply.
To search multiple words/phrases, separate the values with a comma.

Examples:
{COMMAND_PREFIX}sign tiger
{COMMAND_PREFIX}sign ||tiger||
{COMMAND_PREFIX}sign what's up
{COMMAND_PREFIX}sign church, chocolate, computer
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def word_display(word: str, *, template: str = SIGN_TEMPLATE, max_length: int = 100):
    if len(word) > max_length:
        raise errors.BadArgument("⚠️ Input too long. Try a shorter query.")
    quoted_word = quote_plus(word).lower()
    return template.format(
        word_uppercased=word.upper(),
        howsign=f"https://howsign.app/?s={quoted_word}",
        lifeprint=f"https://www.google.com/search?&q=site%3Alifeprint.com+{quoted_word}",
        handspeak=f"https://www.google.com/search?&q=site%3Ahandspeak.com+{quoted_word}",
        signingsavvy=f"https://www.signingsavvy.com/search/{quoted_word}",
        spread_the_sign=f"https://www.spreadthesign.com/en.us/search/?q={quoted_word}",
        youglish=f"https://youglish.com/pronounce/{quoted_word}/signlanguage/asl",
    )


def sign_impl(word: str):
    logger.info(f"sending links for: '{word}'")
    spoiler = get_spoiler_text(word)
    word = spoiler if spoiler else word
    has_multiple = "," in word
    template = SIGN_SPOILER_TEMPLATE if spoiler else SIGN_TEMPLATE
    if has_multiple:
        words = word.split(",")
        embed = discord.Embed()
        for word in words:
            word = word.strip()
            title = f"||{word.upper()}||" if spoiler else word.upper()
            embed.add_field(name=title, value=word_display(word, template=template))
    else:
        title = f"||{word.upper()}||" if spoiler else word.upper()
        embed = discord.Embed(
            title=title,
            description=word_display(word, template=template),
        )

    return {"embed": embed}


HANDSHAPE_HELP = """Show a random or specific handshape

Examples:
{COMMAND_PREFIX}handshape
{COMMAND_PREFIX}handshape claw5

Enter {COMMAND_PREFIX}handshapes to show a list of handshapes.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshape_impl(name: str):
    logger.info(f"handshape: '{name}'")
    try:
        if name == "random":
            handshape = handshapes.get_random_handshape()
        else:
            handshape = handshapes.get_handshape(name)
    except handshapes.HandshapeNotFoundError:
        logger.info(f"handshape '{name}' not found")
        suggestion = did_you_mean(name, tuple(handshapes.HANDSHAPES.keys()))
        if suggestion:
            return {
                "content": f'"{name}" not found. Did you mean "{suggestion}"? Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }
        else:
            return {
                "content": f'"{name}" not found. Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }

    filename = f"{handshape.name}.png"
    file_ = discord.File(handshape.path, filename=filename)
    embed = discord.Embed(title=handshape.name)
    embed.set_image(url=f"attachment://{filename}")
    return {
        "file": file_,
        "embed": embed,
    }


HANDSHAPES_TEMPLATE = """{handshapes}

Enter `{COMMAND_PREFIX}handshape` to display a random handshape or `{COMMAND_PREFIX}handshape [name]` to display a specific handshape.
"""

HANDSHAPES_HELP = """List handshapes

Enter {COMMAND_PREFIX}handshape to display a random handshape or {COMMAND_PREFIX}handshape [name] to display a specific handshape.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshapes_impl():
    return {
        "content": HANDSHAPES_TEMPLATE.format(
            handshapes=", ".join(handshapes.HANDSHAPES.keys()),
            COMMAND_PREFIX=COMMAND_PREFIX,
        )
    }


class ASL(Cog):
    def __init__(self, bot: Bot):
        self.bot = bot

    @command(name="sign", aliases=("howsign", COMMAND_PREFIX), help=SIGN_HELP)
    async def sign_command(self, ctx: Context, *, word: str):
        await ctx.send(**sign_impl(word))

    @sign_command.error
    async def sign_error(self, ctx: Context, error: Exception):
        # Ignore "??"
        if isinstance(error, errors.MissingRequiredArgument):
            if ctx.invoked_with == COMMAND_PREFIX:
                logger.info(
                    f"no argument passed to {COMMAND_PREFIX}{ctx.invoked_with}. ignoring..."
                )
            else:
                await ctx.send(
                    f"⚠️ Enter a word or phrase to search for after `{COMMAND_PREFIX}{ctx.invoked_with}`."
                )

    @command(name="handshape", aliases=("shape",), help=HANDSHAPE_HELP)
    async def handshape_command(self, ctx: Context, name="random"):
        await ctx.send(**handshape_impl(name))

    @command(name="handshapes", aliases=("shapes",), help=HANDSHAPES_HELP)
    async def handshapes_command(self, ctx: Context):
        logger.info("sending handshapes list")
        await ctx.send(**handshapes_impl())


def setup(bot: Bot) -> None:
    bot.add_cog(ASL(bot))
