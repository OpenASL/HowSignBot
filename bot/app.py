import asyncio
import logging

import aiohttp_cors
from aiohttp import web
from ariadne import graphql
from ariadne.constants import PLAYGROUND_HTML

from . import settings
from .bot import bot
from .database import store
from .graphql.schema import schema
from .utils.extensions import walk_extensions

logger = logging.getLogger(__name__)

# Assign app to bot so that extensions can add routes
bot.app = app = web.Application()  # type: ignore

app.cors = cors = aiohttp_cors.setup(
    app,
    defaults={
        "*": aiohttp_cors.ResourceOptions(
            allow_credentials=False,
            expose_headers="*",
            allow_headers="*",
        )
    },
)


async def ping(_):
    return web.Response(body="", status=200)


def graphql_playground(_):
    return web.Response(text=PLAYGROUND_HTML, status=200, content_type="text/html")


async def graphql_server(request):
    data = await request.json()

    success, result = await graphql(
        schema, data, context_value=request, debug=settings.DEBUG
    )

    status = 200 if success else 400
    return web.json_response(result, status=status)


app.add_routes([web.get("/ping", ping)])
resource = bot.app.router.add_resource("/graphql")  # type: ignore
cors.add(resource.add_route("POST", graphql_server))


async def start_bot():
    try:
        await bot.start(settings.DISCORD_TOKEN)
    finally:
        await bot.close()


async def on_startup(app):
    for ext in walk_extensions():
        bot.load_extension(ext)
    await store.connect()
    app["bot_task"] = asyncio.create_task(start_bot())
    app["bot"] = bot


async def on_shutdown(app):
    app["bot_task"].cancel()
    await app["bot_task"]
    await store.disconnect()


app.on_startup.append(on_startup)
app.on_shutdown.append(on_shutdown)
