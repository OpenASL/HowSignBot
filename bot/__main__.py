import logging

from aiohttp import web

from . import __version__, settings
from .app import app

log_format = "%(asctime)s - %(name)s %(levelname)s: %(message)s"
logging.getLogger("discord").setLevel(logging.WARNING)
logging.basicConfig(
    format=log_format,
    level=settings.LOG_LEVEL,
)
logger = logging.getLogger(__name__)

logger.info(f"starting bot version {__version__}")
web.run_app(app, port=settings.PORT)
