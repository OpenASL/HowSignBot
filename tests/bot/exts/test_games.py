import random

import pytest

from bot.exts import games

random.seed(1)


def test_catchphrase(snapshot):
    result = games.catchphrase_impl()
    assert result == snapshot


@pytest.mark.parametrize("category", ("categories", "food", "Hard"))
def test_catchphrase_categories(snapshot, category):
    result = games.catchphrase_impl(category)
    assert result == snapshot
