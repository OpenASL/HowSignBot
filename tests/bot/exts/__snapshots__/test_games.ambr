# name: test_catchphrase
  <class 'dict'> {
    'content': '
      ||softball||
      ||skirt||
      ||green||
      ||bay||
      ||snowman||
      ||Steve Martin||
      ||river rafting||
      ||Spain||
      Categories: easy, medium, hard, animals, food, travel, people, household
    ',
  }
---
# name: test_catchphrase_categories[Hard]
  <class 'dict'> {
    'content': '
      ||audacity||
      ||wreath||
      ||through||
      ||highway||
      ||cardboard||
      ||full moon||
      ||mast||
      ||wool||
      Categories: easy, medium, hard, animals, food, travel, people, household
    ',
  }
---
# name: test_catchphrase_categories[categories]
  <class 'dict'> {
    'content': '
      easy, medium, hard, animals, food, travel, people, household
      Enter `!cp` or `!cp [category]` to generate a list of words/phrases.
    ',
  }
---
# name: test_catchphrase_categories[food]
  <class 'dict'> {
    'content': '
      ||clementine||
      ||sloppy joe||
      ||scone||
      ||chicken pot pie||
      ||blanch||
      ||clams||
      ||Kool-aid||
      ||corn on the cob||
      Categories: easy, medium, hard, animals, food, travel, people, household
    ',
  }
---
